### PR TITLE
Add k8s_exec module to execute command through the API

### PR DESCRIPTION
**Pull request by TristanCacqueray**
##### SUMMARY
This change adds a new module to execute command in a container through the API:
https://docs.okd.io/latest/dev_guide/executing_remote_commands.html#protocol

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
k8s_exec

##### ADDITIONAL INFORMATION
Test playbook:
```yaml
- hosts: localhost
  gather_facts: no
  vars:
    pod: example-sf-zuul-scheduler-84bf645594-kw9lr
    namespace: myproject
  tasks:
    - name: Test Exec
      k8s_exec:
        pod: "{{ pod }}"
        namespace: "{{ namespace }}"
        command: zuul --version

    - name: Test Stderr
      k8s_exec:
        pod: "{{ pod }}"
        namespace: "{{ namespace }}"
        command: sh -c "echo toto > /dev/stderr"
```

Results in:
```paste below
$ ansible-playbook -v test-exec.yaml
PLAY [localhost] *************************************************************

TASK [Test Exec] *************************************************************
changed: [localhost] => {
  "changed": true, "stderr": "", "stderr_lines": [],
  "stdout": "Zuul version: 3.7.2.dev37\n",
  "stdout_lines": ["Zuul version: 3.7.2.dev37"]
}

TASK [Test Stderr] ***********************************************************
changed: [localhost] => {
  "changed": true, "stderr": "toto\n", "stderr_lines": ["toto"],
  "stdout": "", "stdout_lines": []
}

PLAY RECAP *******************************************************************
localhost                  : ok=2    changed=2    unreachable=0    failed=0
```

### CODE REVIEW

**Comment by fabianvf** [comment]
This is really awesome. Out of curiousity, did you consider instead writing a `k8s` connection plugin (similar to the `kubectl` connection plugin but based on python instead of shelling out)? I think generally Ansible prefers to use connection plugins when executing against another host, rather than having modules that open a connection to another host and execute a command. 

This is pretty clean and simple though so I don't have any personal opposition to it, but I'm curious to hear @willthames more expert opinion.

https://docs.ansible.com/ansible/latest/plugins/connection/kubectl.html
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/kubectl.py

**Comment by TristanCacqueray** [comment]
I haven't consider a connection plugin because that exec api is rather limited and it will be hard to use it with generic task like `shell` or `command`. Moreover in the context of the k8s ansible operator, it would be easier to use such task instead of add_host followed by delegated action.

Another improvement to consider is to use a lookup('k8s', kind='Pod') object as a k8s_exec argument.

**Comment by fabianvf** [comment]
Makes sense to me

**Comment by fabianvf** [comment]
shipit

**Comment by fabianvf** [comment]
bot_status

**Comment by geerlingguy** [comment]
Is there any way to get ci re-run on this PR? This would be an extremely-helpful module for my needs, and would save me from having to install `kubectl` alongside Ansible and all its dependencies, or maintaining a custom module to do something similar.

**Comment by TristanCacqueray** [comment]
@geerlingguy alright, I pushed a rebase

**Comment by geerlingguy** [comment]
@TristanCacqueray - I'm trying this module out in my Tower operator, to run a few commands on a Pod... but when I try using it, I get:

```
/usr/lib/python2.7/site-packages/kubernetes/config/kube_config.py:496: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config_dict=yaml.load(f),
Traceback (most recent call last):
  File \"/opt/ansible/.ansible/tmp/ansible-tmp-1573584431.44-267375428717249/AnsiballZ_k8s_exec.py\", line 114, in <module>
    _ansiballz_main()
  File \"/opt/ansible/.ansible/tmp/ansible-tmp-1573584431.44-267375428717249/AnsiballZ_k8s_exec.py\", line 106, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File \"/opt/ansible/.ansible/tmp/ansible-tmp-1573584431.44-267375428717249/AnsiballZ_k8s_exec.py\", line 49, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File \"/tmp/ansible_k8s_exec_payload_YcEo7f/__main__.py\", line 141, in <module>
  File \"/tmp/ansible_k8s_exec_payload_YcEo7f/__main__.py\", line 137, in main
  File \"/tmp/ansible_k8s_exec_payload_YcEo7f/__main__.py\", line 112, in execute_module
  File \"/usr/lib/python2.7/site-packages/kubernetes/stream/stream.py\", line 32, in stream
    return func(*args, **kwargs)
  File \"/usr/lib/python2.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 835, in connect_get_namespaced_pod_exec
    (data) = self.connect_get_namespaced_pod_exec_with_http_info(name, namespace, **kwargs)
  File \"/usr/lib/python2.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 935, in connect_get_namespaced_pod_exec_with_http_info
    collection_formats=collection_formats)
  File \"/usr/lib/python2.7/site-packages/kubernetes/client/api_client.py\", line 321, in call_api
    _return_http_data_only, collection_formats, _preload_content, _request_timeout)
  File \"/usr/lib/python2.7/site-packages/kubernetes/client/api_client.py\", line 155, in __call_api
    _request_timeout=_request_timeout)
  File \"/usr/lib/python2.7/site-packages/kubernetes/stream/stream.py\", line 27, in _intercept_request_call
    return ws_client.websocket_call(config, *args, **kwargs)
  File \"/usr/lib/python2.7/site-packages/kubernetes/stream/ws_client.py\", line 255, in websocket_call
    raise ApiException(status=0, reason=str(e))
kubernetes.client.rest.ApiException: (0)
Reason: Handshake status 403 Forbidden
```

I also did a quick rewrite of the module to make it more in the style of the current `k8s_info` module (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/k8s/k8s_info.py), with all the logic inside `KubernetesExecCommand.execute_module()` instead of inside `main()` (though that didn't help resolve the ApiException). Any idea why this might be happening?

Could it be that I somehow don't have rights to `exec` on Pods? The `kubeconfig` has cluster admin and I'm running it on a local Minikube cluster.

**Review by fabianvf — CHANGES_REQUESTED** [review]
Pods can have multiple containers, so we need a way to specify which container to choose in that case

**Comment by fabianvf** [comment]
@geerlingguy I don't see the same issue, can you verify the same commands work through kubectl?

**Comment by TristanCacqueray** [comment]
@fabianvf not sure why it would fail with 403. I can look into adding container name.

@geerlingguy I don't know how minikube configures the cluster admin, but i can confirm the module works with openshift v3 on fedora. For the record, here is how i tested it:

```shell
$ sudo dnf install -y origin ansible python3-openshift
$ sudo oc cluster up
$ oc login -u devel
$ oc apply -f << EOF
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  restartPolicy: Never
  containers:
    - name: test
      image: fedora
      command: ["/bin/sleep", "Inf"]
EOF

$ # copy the module and write test playbook
$ ansible-playbook -v test.yaml
...
```

**Comment by TristanCacqueray** [comment]
When trying to use k8s_exec with a pod that has 2 container failed with 400 like so:

```
ok: [localhost]
 _________________________________
< TASK [test : Execute a command] >
 ---------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/home/fedora/.ansible/tmp/ansible-tmp-1573587866.1230617-72785839821109/AnsiballZ_k8s_exec.py:18: DeprecationWarning: the imp module is deprecated in favour of importlib;\
 see the module's documentation for alternative uses\n  import imp\nTraceback (most recent call last):\n  File \"/usr/lib/python3.7/site-packages/kubernetes/stream/ws_client.py\", line 253, in websocket_call\n    client = WSClient(configu\
ration, get_websocket_url(url), headers)\n  File \"/usr/lib/python3.7/site-packages/kubernetes/stream/ws_client.py\", line 76, in __init__\n    self.sock.connect(url, header=header)\n  File \"/usr/lib/python3.7/site-packages/websocket/_co\
re.py\", line 223, in connect\n    self.handshake_response = handshake(self.sock, *addrs, **options)\n  File \"/usr/lib/python3.7/site-packages/websocket/_handshake.py\", line 79, in handshake\n    status, resp = _get_resp_headers(sock)\n\
  File \"/usr/lib/python3.7/site-packages/websocket/_handshake.py\", line 152, in _get_resp_headers\n    raise WebSocketBadStatusException(\"Handshake status %d %s\", status, status_message)\nwebsocket._exceptions.WebSocketBadStatusExcept\
ion: Handshake status 400 Bad Request\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/home/fedora/.ansible/tmp/ansible-tmp-1573587866.1230617-72785839821109/AnsiballZ\
_k8s_exec.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/fedora/.ansible/tmp/ansible-tmp-1573587866.1230617-72785839821109/AnsiballZ_k8s_exec.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path,\
 ANSIBALLZ_PARAMS)\n  File \"/home/fedora/.ansible/tmp/ansible-tmp-1573587866.1230617-72785839821109/AnsiballZ_k8s_exec.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/usr/lib64/python3.7\
/imp.py\", line 234, in load_module\n    return load_source(name, filename, file)\n  File \"/usr/lib64/python3.7/imp.py\", line 169, in load_source\n    module = _exec(spec, sys.modules[name])\n  File \"<frozen importlib._bootstrap>\", li\
ne 630, in _exec\n  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/tmp/ansible_k8s_exec_payload_bklz9tjo/__main__.py\"\
, line 125, in <module>\n  File \"/tmp/ansible_k8s_exec_payload_bklz9tjo/__main__.py\", line 112, in main\n  File \"/usr/lib/python3.7/site-packages/kubernetes/stream/stream.py\", line 36, in stream\n    return func(*args, **kwargs)\n  Fi\
le \"/usr/lib/python3.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 835, in connect_get_namespaced_pod_exec\n    (data) = self.connect_get_namespaced_pod_exec_with_http_info(name, namespace, **kwargs)\n  File \"/usr/lib/py\
thon3.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 935, in connect_get_namespaced_pod_exec_with_http_info\n    collection_formats=collection_formats)\n  File \"/usr/lib/python3.7/site-packages/kubernetes/client/api_client\
.py\", line 334, in call_api\n    _return_http_data_only, collection_formats, _preload_content, _request_timeout)\n  File \"/usr/lib/python3.7/site-packages/kubernetes/client/api_client.py\", line 168, in __call_api\n    _request_timeout=\
_request_timeout)\n  File \"/usr/lib/python3.7/site-packages/kubernetes/stream/stream.py\", line 31, in _intercept_request_call\n    return ws_client.websocket_call(config, *args, **kwargs)\n  File \"/usr/lib/python3.7/site-packages/kuber\
netes/stream/ws_client.py\", line 259, in websocket_call\n    raise ApiException(status=0, reason=str(e))\nkubernetes.client.rest.ApiException: (0)\nReason: Handshake status 400 Bad Request\n\n", "module_stdout": "", "msg": "MODULE FAILUR\
E\nSee stdout/stderr for the exact error", "rc": 1}
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by TristanCacqueray** [comment]
With d3eaeb1 you can execute a command on a specific container.

File: lib/ansible/modules/clustering/k8s/k8s_exec.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    type: str
+    required: yes
+  pod:
+    description:
+    - The pod name
+    type: str
+    required: yes
+  container:
+    description:
+    - The container name
```
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_exec.py]: Something like  `- The name of the container in the pod to connect to. Defaults to only container if there is only one container in the pod.` might be a littel clearer, and matches the upstream kubernetes client docs.

**Review by fabianvf — COMMENTED** [review]
Nice, verified the multiple containers are working locally!

File: lib/ansible/modules/clustering/k8s/k8s_exec.py (hunk 1 above)
Comment (TristanCacqueray) [reply lib/ansible/modules/clustering/k8s/k8s_exec.py]: done, thanks for the suggestion.

**Comment by geerlingguy** [comment]
I did what I will relate below, and am now getting the following:

```
      File "/usr/local/lib/python3.7/site-packages/kubernetes/stream/ws_client.py", line 260, in websocket_call
        raise ApiException(status=0, reason=str(e))
    kubernetes.client.rest.ApiException: (0)
    Reason: Handshake status 400 Bad Request
```

1. `minikube delete && minikube start` (to create a fresh cluster)
2. Run the playbook below:

```yaml
- hosts: localhost
  connection: local
  gather_facts: false

  vars:
    ansible_python_interpreter: /usr/local/bin/python3

  tasks:
    - name: Create a test pod.
      k8s:
        state: present
        definition:
          apiVersion: v1
          kind: Pod
          metadata:
            name: test
            namespace: default
          spec:
            restartPolicy: Never
            containers:
              - name: test
                image: fedora
                command: ["/bin/sleep", "Inf"]

    - name: Wait for pod to enter running state.
      k8s_info:
        kind: Pod
        name: test
        namespace: default
      register: test_pod
      until: "test_pod['resources'][0]['status']['phase'] == 'Running'"
      retries: 5
      delay: 5

    - name: Execute a command on the test pod.
      k8s_exec:
        pod: test
        namespace: default
        command: date
      register: exec_result

    - debug: var=exec_result
```

It fails on the `k8s_exec` task... not sure why though.

**Comment by geerlingguy** [comment]
Hmm... I wonder if it has to do with this: https://stackoverflow.com/a/54051435/100134

> I believe you're using minikube for development purpose. It is not able to recognise your hostname. You can make it work by disabling assert_hostname in your program like:
> 
>     from kubernetes.client import configuration 
>     config.load_kube_config()                                                 
>     configuration.assert_hostname = False
> 
> This should resolve your issue.

**Comment by geerlingguy** [comment]
Building a minimal script to run the command directly using the same python:

```python
from pprint import pprint

from kubernetes import client, config
from kubernetes.client import CoreV1Api
from kubernetes.stream import stream

config.load_kube_config()

exec_command = ['date']

api = CoreV1Api()
api_response = stream(api.connect_get_namespaced_pod_exec,
                      'test',
                      'default',
                      command=exec_command,
                      stderr=True,
                      stdin=False,
                      stdout=True,
                      tty=False,
                      _preload_content=True)

pprint(api_response)
```

It works without issue, giving:

```
$ python3 test.py
'Tue Nov 12 21:51:26 UTC 2019\n'
```

**Comment by geerlingguy** [comment]
It looks like dropping the `container=module.params.get("container"),` line results in a working test playbook for me.

**Comment by fabianvf** [comment]
@TristanCacqueray it looks like providing `None` as the container argument breaks things (bug in the codegen likely). Adding the container argument like this worked fine, even if it's a bit janky:

```
diff --git a/lib/ansible/modules/clustering/k8s/k8s_exec.py b/lib/ansible/modules/clustering/k8s/k8s_exec.py
index 496f862afd..01c5f350b9 100644
--- a/lib/ansible/modules/clustering/k8s/k8s_exec.py
+++ b/lib/ansible/modules/clustering/k8s/k8s_exec.py
@@ -124,17 +124,21 @@ def main():
     # Load kubernetes.client.Configuration
     module.get_api_client()
     api = core_v1_api.CoreV1Api()
+
+    # hack because passing the container as None breaks things
+    optional_kwargs = {}
+    if module.params.get('container'):
+        optional_kwargs['container'] = module.params['container']
     resp = stream(
         api.connect_get_namespaced_pod_exec,
         module.params["pod"],
         module.params["namespace"],
-        container=module.params.get("container"),
         command=shlex.split(module.params["command"]),
         stdout=True,
         stderr=True,
         stdin=False,
         tty=False,
-        _preload_content=False)
+        _preload_content=False, **optional_kwargs)
     stdout, stderr = [], []
     while resp.is_open():
         resp.update(timeout=1)
```

**Comment by geerlingguy** [comment]
Dropping that makes the module work in the context of a local connection running on my Mac into the Minikube cluster... but if I'm inside the cluster, and run the module on the Ansible Operator image I'm working on, it still results in the `Reason: Handshake status 403 Forbidden` :(

**Comment by TristanCacqueray** [comment]
@fabianvf thanks for the fix, next commit will include it.

@geerlingguy it seems like your user is unable to use the exec api. Here is what openshift configure by default:

```
$ oc describe rolebinding admin
Name:         admin
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  ClusterRole
  Name:  admin
Subjects:
  Kind  Name   Namespace
  ----  ----   ---------
  User  devel  

$ oc get clusterrole admin -o yaml
aggregationRule:
  clusterRoleSelectors:
  - matchLabels:
      rbac.authorization.k8s.io/aggregate-to-admin: "true"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    openshift.io/description: A user that has edit rights within the project and can
      change the project's membership.
    rbac.authorization.kubernetes.io/autoupdate: "true"
  creationTimestamp: 2019-11-12T19:28:17Z
  labels:
    kubernetes.io/bootstrapping: rbac-defaults
  name: admin
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - pods/attach
  - pods/exec
  - pods/portforward
  - pods/proxy
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch

```

**Comment by TristanCacqueray** [comment]
hum, the [failed job](https://app.shippable.com/github/ansible/ansible/runs/150298/35/console) doesn't seem related (`Failed to create vault token for this job.`)

**Comment by geerlingguy** [comment]
The last commits fixed the issue for me in my standalone test playbook. I'm checking my operator's RBAC and it seems like for the pod in question (which it's trying to exec on), it should have access:

```
- apiGroups:
  - tower.ansible.com
  resources:
  - '*'
  verbs:
  - '*'
```

But it is probably not related to this module's code, so the module itself gets the green light from me. Test failure looks unrelated and probably just needs a re-run.

**Comment by TristanCacqueray** [comment]
@geerlingguy thank you for the feedback. About your RBAC, iiuc the pods exec api is in the v1 apiGroups, or the rule could use: `apiGroups: [""]`.

**Comment by geerlingguy** [comment]
@TristanCacqueray ah, thanks. Digging further, I found I had to change my rule for pods to be like so:

```
- apiGroups:
  - ""
  resources:
  - pods/exec
  verbs:
  - create
  - get
```

Found in this comment: https://github.com/kubernetes-client/python/issues/690#issuecomment-459010364

Strangely, though, I'm now getting the following error from `stream.py`, which doesn't make sense to me:

```
File \"/usr/lib/python2.7/site-packages/kubernetes/stream/ws_client.py\", line 255, in websocket_call
    raise ApiException(status=0, reason=str(e))
kubernetes.client.rest.ApiException: (0)
Reason: Handshake status 200 OK
```

**Comment by TristanCacqueray** [comment]
Something is broken indeed, I now also get `kubernetes.client.rest.ApiException: (0)` when using this module within an operator running on okd.

**Comment by fabianvf** [comment]
@TristanCacqueray I think that's an issue with the proxy that the Ansible Operator creates, it seems for whatever reason it is unable to handle the `exec` connections made by the Python client, although it doesn't seem to have trouble with the `exec` implementation that `kubectl` uses. We still haven't figured out what the cause might be but I think it lies in the Python client or the proxy, not in your module

**Comment by TristanCacqueray** [comment]
@fabianvf that would make sense. IIRC the kubectl clients uses the legacy `spdy` protocol while the python client uses the new `websocket` protocol. Would it be possible to shortcut the operator proxy for that module?

**Comment by geerlingguy** [comment]
@TristanCacqueray - I can confirm that it's an issue with the Operator proxy as @fabianvf states; see https://github.com/operator-framework/operator-sdk/issues/2204 for the downstream bug report.

**Comment by geerlingguy** [comment]
@TristanCacqueray - The Kubernetes-related modules and plugins have been moved to a new `community.kubernetes` collection, where further development and module additions (like this one) will go. Would you be willing/able to create a new PR adding this module to that project?

Collection repo: https://github.com/ansible-collections/kubernetes

As I've already been testing this module thoroughly related to my work on some operators (it is extremely useful for many situations, and will be for more once that operator-sdk proxying issue is resolved, I for one am +1 to shipping it quickly once submitted!

**Comment by TristanCacqueray** [comment]
@geerlingguy sure thing, here is the initial draft: https://github.com/ansible-collections/kubernetes/pull/14

**Comment by TristanCacqueray** [comment]
This now lives in https://github.com/ansible-collections/kubernetes/pull/14

**Pull request closed without merging**